Adding Replay mode

Cealgair opened the pull request:
The Replay mode would allows you to import a "game state" json file and replay the actions step by step, instead of rushing them all at once like when you just import the game state.

At the time of opening this draft pull request it's still very broken, mainly because it does not stop you from inserting new actions in the middle of the replay, thus screwing things up significantly.

Cealgair committed:
Very crude replay mode

Cealgair committed:
Locked most actions in replay mode

Cealgair commented:
Still probably missing a few quality of life features, but at least I think it should not be broken now

xxmichaellong committed:
Merge branch 'main' into replay

SudeshS commented:
God tier PR

Eyoel-gebre commented:
banger

xxmichaellong commented:
Hi Cealgair!

Thank you so much for working on this PR. I really appreciate the initiative and I'm excited to add this in ASAP :).

Here were a couple issues that I ran into: 

- importing a replay doesn’t work when you do it multiple times; i.e., when you import a replay, exit replay, and then try to import another replay, it doesn't work
- if importing replay fails, the sim doesn’t seem to go completely back to normal; e.g., keybinds are not turned back on
- i think the “fast forward” and “rewind” buttons could have better names. maybe just “start” and “end”?. also, the color for “next” and “prev” buttons shouldn’t change when you flip the board. would probably just keep it gray.
- clicking “prev” and waiting for it to revert one action back can take a while, especially when it’s deep in a game (similar issue to the undo button) - this means a player may click it multiple times thinking that it didn’t work the first time and thus queuing multiple undos; it would be good to not run “prev” until the current “prev” has finished running. maybe use a promise?
- the import replay button should be hidden in 2P (and exit replay button, if the player was in a replay and then joined 2P) 
- you should not be able to import decks during replay mode.
- we should hide buttons that can’t be used in replay mode (i.e., shuffle button when you click on the deck, all of the buttons that appear when you are looking at cards from the deck (see photo below)
<img width="1487" alt="Pasted Graphic" src="https://github.com/user-attachments/assets/f3369e52-9374-4e32-9eb0-4d18615ad849">



lastly, something to think about: i’m not sure if we need a “import replay” and “import game state”? i guess import game state is useful when you want to continue playing from that position (works in p1 and p2), whereas replay is more for rewatching what’s been played (although you are able to exit at any time and continue playing from there). or maybe the names could be better so it’s clear what each one is for.

Let me know if you have any questions, and I'm also happy to take on some of these bugs if you let me know which ones!

xxmichaellong commented on src/js/front-end/actions/keybinds/keybinds.js:
```diff
@@ -78,6 +82,9 @@ export const keyDown = (event) => {
             shuffleBoard(systemState.initiator, systemState.initiator);
         };
         if (event.key === 'f' && !(event.altKey || event.getModifierState('Alt'))) {
+            if(systemState.isReplay && !systemState.isTwoPlayer){
```
instead of writing this a bunch of times for all the respective keybinds, let's put all of the relevant keybinds nested in one if statement (or move all of the ones that aren't impacted to the top, and then afterwards have an if statement that returns early)

Cealgair replied:
I moved the unaffected ones to the top in [9588910](https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/9588910523e901a643f28aa12aad8b81cf303b6f). Had to split an if statement to do that, but I thought it was the cleanest way possible

xxmichaellong commented on src/js/front-end/actions/zones/general.js:
```diff
@@ -10,8 +10,12 @@ import { addAbilityCounter } from "../counters/ability-counter.js";
 
-export const shuffleAll = (user, initiator, zoneId, indices, emit = true) => {
+export const shuffleAll = (user, initiator, zoneId, indices, emit = true, isFromReplay = false) => {
+    if (replayBlock('action','shuffleAll',isFromReplay)){
```
i think the buttons and keybinds that enable shuffleAll, shuffleBottom etc. should be hidden/disabled during replay mode. then, you don't need to do this check for all of the functions (since the user should have no way to trigger it in replay mode)

Cealgair replied:
Done in [91bbd5e](https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/91bbd5ef3e30bfa44b35e3eb347ad2ad2d770186), I left the VSTAR and GX buttons since they also serve as indicators

xxmichaellong commented on src/js/front-end/setup/general/accept-action.js:
```diff
@@ -84,11 +85,24 @@ const actionToFunction = (action) => {
+        }
     } else {
-        actionToFunction(action)(user, emit);
+        if(['shuffleAll','shuffleBottom','discardAll','lostZoneAll','handAll','leaveAll','VSTARGXFunction'].includes(action)){
```
i think you'll be able to remove this check once the buttons/keybinds for all of these functions are hidden/disabled.

Cealgair replied:
Done in [83be7ec](https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/83be7eca0da088e1045be0ea3836abff653fd95b), although again I left the check for VSTAR and GX

xxmichaellong commented on src/js/front-end/setup/general/replay-block.js:
```diff
@@ -0,0 +1,11 @@
+    "contextMenu": ['lookPrizesButton','revealHidePrizesButton','lookHandButton','prizesHeader','handHeader','deckHeader','boardHeader'],
+};
+
+export const replayBlock = (type, value, isFromReplay = false) => {
```
i think i'd rename this to `isActionBlockedByReplay` for clarity.

xxmichaellong replied:
or even just `isBlockedByReplay`

Cealgair replied:
Done in [6621d23](https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/6621d237b246d021b4310c791c6d1249b9379059)

Cealgair committed:
reduced isReplay checks in keybinds.js

Cealgair committed:
replay mode: renamed buttons, locked deck import, handled failed import

Cealgair committed:
Merge branch 'main' into replay

Cealgair committed:
replay mode: locked 2P, renamed 'import replay' to 'enter replay mode'

Cealgair committed:
replay mode: mitigated prev spam bug, hidden keybindReminder, resized top buttons

Cealgair commented:
> clicking “prev” and waiting for it to revert one action back can take a while, especially when it’s deep in a game (similar issue to the undo button) - this means a player may click it multiple times thinking that it didn’t work the first time and thus queuing multiple undos; it would be good to not run “prev” until the current “prev” has finished running. maybe use a promise?

Like I said earlier on Discord, I was trying to make it so if the user presses "prev" while another "prev" is running, then nothing happens. I did not manage to do that, however I did stumble upon multiple ways to make it so if the user does that then ptcgsim waits until the current "prev" has finished and then runs another "prev". One such way is what I implemented in [a3a457d](https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/a3a457df1e3a74caf973fec1d2bb003dd2796708).

> importing a replay doesn’t work when you do it multiple times; i.e., when you import a replay, exit replay, and then try to import another replay, it doesn't work

Also like I said in the Discord, I have no idea what is causing this. So for these two issues I think I need help

> the color for “next” and “prev” buttons shouldn’t change when you flip the board. would probably just keep it gray.

After thinking about this, I think I disagree: those two are the buttons the user will want to click on the most, so I think it's sensible to give them the highlight color. I would also be happy with keeping just Next with the color and make Prev gray

Cealgair committed:
replay mode: hidden options in 2P

Cealgair committed:
renamed replayBlock to isBlockedByReplay

Cealgair committed:
Hidden locked buttons from replay mode

Cealgair committed:
removed some isFromReplay checks

Cealgair commented:
I think I've done everything except the three things I listed in my previous comment

Cealgair committed:
locked user from re-entering replay mode

Cealgair commented:
Since I couldn't find the bug that breaks everything when trying to re-enter replay mode, I just made it impossible to re-enter with [39b960c](https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/39b960c4743be995b33a7808efe3aa817ce1475a). Of course feel free to revert that commit if the bug gets actually fixed

xxmichaellong committed:
fix keybinds in replay mode

xxmichaellong committed:
clean up header logic

xxmichaellong committed:
clean up vstargx code

xxmichaellong committed:
fix colors and icons

xxmichaellong commented:
> Since I couldn't find the bug that breaks everything when trying to re-enter replay mode, I just made it impossible to re-enter with [39b960c](https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/39b960c4743be995b33a7808efe3aa817ce1475a). Of course feel free to revert that commit if the bug gets actually fixed

hey, got a chance to look at it today. just wanted to update you that i am in progress of reviewing the code-have sent a few docs for nitpicky things. hopefully tomorrow i can figure out the 3 blockers you mentioned that are stopping us from shipping this.

thanks again!

Cealgair committed:
fixed socket

Cealgair commented:
I saw you unlocked the shift keybind in [3fe8b66](https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/3fe8b6691e12ff29ffbe03b9bb273627137875d4), I'm not sure it's a good idea to display the list of all keybinds while most of them are disabled during replay mode. If you still think it's better to still show them, I guess keybindReminder should not be hidden during replay mode, so that change from [a3a457d](https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/a3a457df1e3a74caf973fec1d2bb003dd2796708) would need to be reverted

Cealgair committed:
Merge branch 'xxmichaellong:main' into replay

Cealgair committed:
Merge branch 'main' into replay

Cealgair committed:
Fixed broken merge

Cealgair commented:
I just noticed that the drag functions do not get re-enabled when exiting replay mode (although the keybinds do). I will look into it when I have the time

EDIT: the options in the context menu that appears when right-clicking on a card also don't get re-enabled

EDIT 2: the coin flip keybind does not get re-enabled, but the button on the field does

EDIT 3: the VSTAR/GX buttons on the field do not get re-enabled

Cealgair committed:
exitReplayMode now resets systemState.isReplay

Cealgair commented:
> I just noticed that the drag functions do not get re-enabled when exiting replay mode (although the keybinds do). I will look into it when I have the time
> 
> EDIT: the options in the context menu that appears when right-clicking on a card also don't get re-enabled
> 
> EDIT 2: the coin flip keybind does not get re-enabled, but the button on the field does
> 
> EDIT 3: the VSTAR/GX buttons on the field do not get re-enabled

Well, it was a super simple fix, see https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/66b33405b72cad1b9a38c84b9f4bdf74608a6a0c

EDIT: also fixed plenty of other issues, so I also made https://github.com/xxmichaellong/ptcg-sim/pull/19/commits/2b0dd00d1d517a7b04484adb76f0d7db1c7c6192

EDIT 2: I reverted the comment I mentioned in the previous edit, cause you still can't always re-enter replay mode after exiting. Not sure what I tested when I made that commit, but anyway the sim will stil break sometimes

Cealgair committed:
Replay mode is no longer locked after exiting

Cealgair committed:
reverted 2b0dd00d1d517a7b04484adb76f0d7db1c7c6192

xxmichaellong committed:
undo fix, replay enter/exit fix

xxmichaellong committed:
return early if no file

xxmichaellong committed:
change socket url

xxmichaellong committed:
fix undo

xxmichaellong committed:
undo async not exported

xxmichaellong committed:
fix undo in replay

xxmichaellong reviewed (approved):
Finally got around to finishing this PR. Thank you @Cealgair for all of the hard work. Happy to get this over the finish line before the end of the year.

The one unaddressed issue, which I also think is out of the scope of this PR, is the issue of multiple "undos" or "rewinds" queuing up, which causes lag and sometimes even crashes the game. I was not able to figure out a simple solution unfortunately, but I don't think that it should stop this PR from being shipped.

Feel free to PM me on Discord or comment on this PR with any comments/bug finds. I will say that I could have more thoroughly tested the replay, but I feel pretty good about it.

Hope you are having a happy holidays!!!

-xxl
